File: src/db-toolkit-ui/resources/backend/_internal/operations/ai/schema_analyzer.py
"""Schema Analyzer for DBAssist AI functionality."""

from utils.logger import logger
from typing import Dict, List, Any
from core.config import settings
from .cloudflare_client import CloudflareAIClient
from .prompts import SCHEMA_ANALYSIS_PROMPT, SCHEMA_RECOMMENDATIONS_PROMPT, SYSTEM_PROMPT


class SchemaAnalyzer:
    """AI assistant for database schema analysis."""

    def __init__(self):
        """Initialize schema analyzer."""
        if not settings.has_cloudflare_credentials:
            raise ValueError("Cloudflare credentials not configured")
        
        self.client = CloudflareAIClient(
            account_id=settings.cloudflare_account_id,
            api_token=settings.cloudflare_api_token
        )
        self.model = settings.cloudflare_model
        self.temperature = settings.ai_temperature
        self.max_tokens = settings.ai_max_tokens

    async def analyze_table(
        self, 
        table_name: str, 
        columns: List[Dict], 
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Analyze a database table and provide insights."""
        try:
            # Format schema info
            schema_info = f"Table: {table_name}\n"
            schema_info += "Columns:\n"
            for col in columns:
                schema_info += f"  - {col.get('name')} ({col.get('type', 'unknown')})\n"
            
            prompt = SCHEMA_ANALYSIS_PROMPT.format(
                db_type=db_type,
                schema_info=schema_info
            )
            
            response = await self.client.generate(
                prompt=prompt,
                model=self.model,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                system_prompt=SYSTEM_PROMPT
            )
            
            return {
                "success": True,
                "table_name": table_name,
                "summary": response,
                "purpose": "Data storage and retrieval",
                "suggestions": [],
                "common_queries": [],
                "index_recommendations": [],
                "relationship_suggestions": []
            }
        except Exception as e:
            logger.error(f"AI operation error: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "table_name": table_name,
                "summary": "",
                "suggestions": []
            }

    async def suggest_indexes(
        self, 
        table_name: str, 
        columns: List[Dict],
        query_patterns: List[str] = None,
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Suggest database indexes for better performance."""
        try:
            # Basic index suggestions based on column types and patterns
            suggestions = []
            
            for column in columns:
                col_name = column.get('name', '')
                col_type = column.get('type', '').lower()
                
                # Suggest indexes for common patterns
                if 'id' in col_name.lower() and col_name != 'id':
                    suggestions.append({
                        "type": "btree",
                        "columns": [col_name],
                        "reason": f"Foreign key column {col_name} would benefit from an index"
                    })
                elif col_type in ['timestamp', 'date', 'datetime']:
                    suggestions.append({
                        "type": "btree", 
                        "columns": [col_name],
                        "reason": f"Date/time column {col_name} commonly used in WHERE clauses"
                    })
                elif 'email' in col_name.lower():
                    suggestions.append({
                        "type": "btree",
                        "columns": [col_name],
                        "reason": f"Email column {col_name} likely used for lookups"
                    })

            return {
                "success": True,
                "table_name": table_name,
                "index_suggestions": suggestions,
                "performance_impact": "medium"
            }
        except Exception as e:
            logger.error(f"AI operation error: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "index_suggestions": []
            }

    async def analyze_relationships(
        self, 
        tables: Dict[str, List[Dict]], 
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Analyze potential relationships between tables."""
        try:
            relationships = []
            table_names = list(tables.keys())
            
            # Look for potential foreign key relationships
            for table_name, columns in tables.items():
                for column in columns:
                    col_name = column.get('name', '').lower()
                    
                    # Check if column name suggests a foreign key
                    if col_name.endswith('_id') and col_name != 'id':
                        referenced_table = col_name[:-3]  # Remove '_id'
                        
                        # Check if referenced table exists
                        for other_table in table_names:
                            if other_table.lower() == referenced_table or other_table.lower() == referenced_table + 's':
                                relationships.append({
                                    "from_table": table_name,
                                    "from_column": column.get('name'),
                                    "to_table": other_table,
                                    "to_column": "id",
                                    "relationship_type": "many_to_one",
                                    "confidence": "medium"
                                })

            return {
                "success": True,
                "relationships": relationships,
                "total_suggestions": len(relationships)
            }
        except Exception as e:
            logger.error(f"AI operation error: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "relationships": []
            }

    async def generate_common_queries(
        self, 
        table_name: str, 
        columns: List[Dict],
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Generate common SQL queries for a table."""
        try:
            queries = []
            col_names = [col.get('name') for col in columns if col.get('name')]
            
            # Basic SELECT query
            queries.append({
                "name": "Select All Records",
                "query": f"SELECT * FROM {table_name} LIMIT 10;",
                "description": "Retrieve first 10 records from the table"
            })
            
            # Count query
            queries.append({
                "name": "Count Records",
                "query": f"SELECT COUNT(*) FROM {table_name};",
                "description": "Count total number of records"
            })
            
            # Find queries based on column types
            for column in columns:
                col_name = column.get('name', '')
                col_type = column.get('type', '').lower()
                
                if 'timestamp' in col_type or 'date' in col_type:
                    queries.append({
                        "name": f"Recent Records by {col_name}",
                        "query": f"SELECT * FROM {table_name} WHERE {col_name} >= NOW() - INTERVAL '7 days' ORDER BY {col_name} DESC;",
                        "description": f"Find records from the last 7 days based on {col_name}"
                    })
                elif 'email' in col_name.lower():
                    queries.append({
                        "name": f"Find by {col_name}",
                        "query": f"SELECT * FROM {table_name} WHERE {col_name} = 'example@email.com';",
                        "description": f"Find records by {col_name}"
                    })

            return {
                "success": True,
                "table_name": table_name,
                "queries": queries[:5],  # Limit to 5 queries
                "total_generated": len(queries)
            }
        except Exception as e:
            logger.error(f"AI operation error: {str(e)}")
            return {
                "success": False,
                "error": str(e),
                "queries": []
            }


# Global instance
schema_analyzer = SchemaAnalyzer()

File: src/db-toolkit-ui/electron/backend/operations/ai/schema-analyzer.js
/**
 * Schema Analyzer for DBAssist AI functionality.
 */

const { CloudflareAIClient } = require('./cloudflare-client');
const { SCHEMA_ANALYSIS_PROMPT, SYSTEM_PROMPT } = require('./prompts');

class SchemaAnalyzer {
  constructor(accountId, apiToken, model = '@cf/meta/llama-3.1-70b-instruct', temperature = 0.7, maxTokens = 2048) {
    if (!accountId || !apiToken) {
      throw new Error('Cloudflare credentials not configured');
    }
    
    this.client = new CloudflareAIClient(accountId, apiToken);
    this.model = model;
    this.temperature = temperature;
    this.maxTokens = maxTokens;
  }

  async analyzeTable(tableName, columns, dbType = 'postgresql') {
    try {
      let schemaInfo = `Table: ${tableName}\n`;
      schemaInfo += 'Columns:\n';
      for (const col of columns) {
        schemaInfo += `  - ${col.name} (${col.type || 'unknown'})\n`;
      }
      
      const prompt = SCHEMA_ANALYSIS_PROMPT
        .replace('{db_type}', dbType)
        .replace('{schema_info}', schemaInfo);
      
      const response = await this.client.generate(
        prompt,
        this.model,
        this.maxTokens,
        this.temperature,
        SYSTEM_PROMPT
      );
      
      return {
        success: true,
        table_name: tableName,
        summary: response,
        purpose: 'Data storage and retrieval',
        suggestions: [],
        common_queries: [],
        index_recommendations: [],
        relationship_suggestions: []
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        table_name: tableName,
        summary: '',
        suggestions: []
      };
    }
  }

  async suggestIndexes(tableName, columns, queryPatterns = null, dbType = 'postgresql') {
    try {
      const suggestions = [];
      
      for (const column of columns) {
        const colName = column.name || '';
        const colType = (column.type || '').toLowerCase();
        
        if (colName.toLowerCase().includes('id') && colName !== 'id') {
          suggestions.push({
            type: 'btree',
            columns: [colName],
            reason: `Foreign key column ${colName} would benefit from an index`
          });
        } else if (colType.includes('timestamp') || colType.includes('date') || colType.includes('datetime')) {
          suggestions.push({
            type: 'btree',
            columns: [colName],
            reason: `Date/time column ${colName} commonly used in WHERE clauses`
          });
        } else if (colName.toLowerCase().includes('email')) {
          suggestions.push({
            type: 'btree',
            columns: [colName],
            reason: `Email column ${colName} likely used for lookups`
          });
        }
      }

      return {
        success: true,
        table_name: tableName,
        index_suggestions: suggestions,
        performance_impact: 'medium'
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        index_suggestions: []
      };
    }
  }

  async analyzeRelationships(tables, dbType = 'postgresql') {
    try {
      const relationships = [];
      const tableNames = Object.keys(tables);
      
      for (const [tableName, columns] of Object.entries(tables)) {
        for (const column of columns) {
          const colName = (column.name || '').toLowerCase();
          
          if (colName.endsWith('_id') && colName !== 'id') {
            const referencedTable = colName.slice(0, -3);
            
            for (const otherTable of tableNames) {
              if (otherTable.toLowerCase() === referencedTable || 
                  otherTable.toLowerCase() === referencedTable + 's') {
                relationships.push({
                  from_table: tableName,
                  from_column: column.name,
                  to_table: otherTable,
                  to_column: 'id',
                  relationship_type: 'many_to_one',
                  confidence: 'medium'
                });
              }
            }
          }
        }
      }

      return {
        success: true,
        relationships,
        total_suggestions: relationships.length
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        relationships: []
      };
    }
  }

  async generateCommonQueries(tableName, columns, dbType = 'postgresql') {
    try {
      const queries = [];
      
      queries.push({
        name: 'Select All Records',
        query: `SELECT * FROM ${tableName} LIMIT 10;`,
        description: 'Retrieve first 10 records from the table'
      });
      
      queries.push({
        name: 'Count Records',
        query: `SELECT COUNT(*) FROM ${tableName};`,
        description: 'Count total number of records'
      });
      
      for (const column of columns) {
        const colName = column.name || '';
        const colType = (column.type || '').toLowerCase();
        
        if (colType.includes('timestamp') || colType.includes('date')) {
          queries.push({
            name: `Recent Records by ${colName}`,
            query: `SELECT * FROM ${tableName} WHERE ${colName} >= NOW() - INTERVAL '7 days' ORDER BY ${colName} DESC;`,
            description: `Find records from the last 7 days based on ${colName}`
          });
        } else if (colName.toLowerCase().includes('email')) {
          queries.push({
            name: `Find by ${colName}`,
            query: `SELECT * FROM ${tableName} WHERE ${colName} = 'example@email.com';`,
            description: `Find records by ${colName}`
          });
        }
      }

      return {
        success: true,
        table_name: tableName,
        queries: queries.slice(0, 5),
        total_generated: queries.length
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        queries: []
      };
    }
  }
}

let _schemaAnalyzer = null;

function getSchemaAnalyzer() {
  if (_schemaAnalyzer === null) {
    try {
      const accountId = process.env.CLOUDFLARE_ACCOUNT_ID;
      const apiToken = process.env.CLOUDFLARE_API_TOKEN;
      
      if (!accountId || !apiToken) {
        return null;
      }
      
      _schemaAnalyzer = new SchemaAnalyzer(accountId, apiToken);
    } catch (error) {
      return null;
    }
  }
  return _schemaAnalyzer;
}

module.exports = { SchemaAnalyzer, getSchemaAnalyzer };
